### ISSUE TITLE
mksnapshot-v1.6.0-linux-armv7l.zip contains x86 binary not armv7l binary

### ISSUE THREAD

**Issue by mathuin** [electron/electron#11413]
* Electron version: v1.6.0
* Operating system: armv7l

### Expected behavior

I downloaded https://github.com/electron/electron/releases/download/v1.6.0/mksnapshot-v1.6.0-linux-armv7l.zip and it should have contained an armv7l binary.

### Actual behavior

Instead it contained an x86 binary.

### How to reproduce

Click the URL, unzip the zip file, and run `file mksnapshot`.  If it says "Intel 80386", that's bad, because it's supposed to say "ARM" instead.  This problem is keeping me from building Atom on my Chromebook because I can't find an ARM binary for `mksnapshot`.

**Comment by shreyneil** [comment]
A similar issue has been raised previously  

https://github.com/electron/mksnapshot/issues/3
and https://github.com/hypersad found a way for building 
atom. 
Here is the link to it.
https://github.com/hypersad/atom-armv7l
Please verify and get back.

**Comment by mathuin** [comment]
Yes, that is the same issue that I am facing.  I honestly wasn't sure where to post about it, but it's affecting me for the same reason it affected them.  

My first attempt to build from that user's repository (marked unmaintained last month, by the way) results in a failure with code ELIFECYCLE for `chromedriver-v2.21-linux-arm.zip`.  The release page for electron/chromedriver only goes up to v1.7.1 and I don't know where else it might be looking.  That way lies pain, especially because I do not know anything about node.

What can be done about the root cause of this issue?  Is there a way to fix the problem of releasing binaries built for x86 in a tar file labeled armv7l ?

**Comment by groundwater** [comment]
Going a bit off memory of also trying to get Atom running on ARM64.

I believe `mksnapshot` does not work on ARM64. You have to compile it for x86/64 essentially like a cross-compiler, and the x86 version outputs the ARM64 snapshot. We never got this far, and just disabled snapshotting altogether.

As to whether the bundled x86 version actually does this is another story. I don't think it's tested, and CI has been outputting it as an artifact because that's how the scripts work.

**Comment by mathuin** [comment]
If I could successfully cross-compile Atom itself for armv7l on x86_64, that would be awesome.  It's hard to find information about that because of the other relevant hits for cross-compile and Atom, and there's no info in the Flight Manual about that.

With regard to disabling snapshotting, it's still enabled in the master branch as far as I can tell.  `script/lib/generate-startup-snapshot.js` is present and is executed (or at least its output is seen) when I try to build Atom.  If there's an elegant way for me to disable that so I can skip this step, please let me know!

**Comment by hypersad** [comment]
I'm so lazy, but if new correct compiled mksnapshot binaries will be released, I could try to find my old laptop, install xubuntu on it again and try to build new Atom.

**Comment by mathuin** [comment]
If new correct compiled mksnapshot binaries are released, then the branch containing https://github.com/atom/atom/pull/15431 should be able to build Atom on ARM as well.  In theory. :-)

**Comment by hypersad** [comment]
@mathuin the main problem, as I remember, was incorrect arm architectute detection, currently I'm have an idea how I can fix that.

**Comment by nonnymoose** [comment]
I can confirm that @groundwater is correct.
I downloaded the mksnapshot armv7l binary and ran `mksnapshot --help` on my x86 machine, and here are the first few lines:

    target simulator arm v7 neon hard
    ARMv8=1 ARMv7=1 VFPv3=1 VFP32DREGS=1 NEON=1 SUDIV=1 USE_EABI_HARDFLOAT=1

It might be possible to run this on the host using qemu-user.

**Comment by hypersad** [comment]
I don't think that is right to set the `version/1.6x` because this probably affects all prebuilt mksnapshot binaries.

**Comment by groundwater** [comment]
Thanks for checking @nonnymoose.

Unless something has changed in recent Chromium, tooling will need to run `mksnapshot` from an x86-64 host machine to generate and incorporate the ARM64 snapshots.

I believe the x86 binary for `mksnapshot` is working as intended.

**Comment by hypersad** [comment]
@groundwater so, you mean we can use x86_64 `mksnapshot` to generate correct startup blobs from startup scripts?

**Comment by groundwater** [comment]
@hypersad I believe so, the x86_64 binaries _should_ in theory generate the correct ARM7/ARM8 snapshots. We stumbled across this when trying to get Atom running. Everything ARM related in Chromium seems to be cross-compiled, so it makes sense they might design it this way.

Unfortunately I have not tested this, and if the generated snapshots do not work, that's definitely a valid bug, but we should perhaps deal with that in a new issue.

**Comment by hypersad** [comment]
@groundwater I'm currently trying to build `v8` on my Raspberry Pi 3 Model B. I've also tried to cross-compile it and end with success, but there was NO `mksnapshot` binary, lol.

Yesterday I tried to use ia32 `mksnapshot` on my Windows notebook and got  `startup_blob.bin`, but Atom throw `Ilegal instruction` exception into me :(

**Comment by groundwater** [comment]
> but there was NO mksnapshot binary, lol

@hypersad my recollection is that compiling chromium on ARM64 is not really supported, and that you cannot cross-compile `mksnapshot`. However `mksnapshot` can be compiled to itself be a cross-compiler. I believe the `gn` build tool will set all this up correctly, but the process is a little groggy. I wonder if there is also a difference between gyp and gn here. We compile v8 with libchromiumcontent which uses gn. I think most people who compile v8 in isolation use gyp. I have no insight into how up to date or in sync the two methods are.

We documented our process here https://github.com/atom/atom/issues/15881 although ultimately we just removed snapshotting to get our prototype working. It's something we need to loop back on.

**Comment by hypersad** [comment]
@groundwater I've got only blank Atom window on armv7l without snapshotting. It can't find main module.

**Comment by hypersad** [comment]
Got it!
`CXX(target) /home/pi/Coding/v8/out/arm.release/obj.target/mksnapshot/src/snapshot/mksnapshot.o
LINK(target) /home/pi/Coding/v8/out/arm.release/mksnapshot`
`pi@raspberrypi:~/Coding/v8/out/arm.release $ file mksnapshot
mksnapshot: ELF 32-bit LSB executable, ARM, EABI5 version 1 (GNU/Linux), dynamically linked, interpreter /lib/ld-linux-armhf.so.3, for GNU/Linux 3.2.0, BuildID[sha1]=78106edf61705ebf79d8206966383867009d5bb6, not stripped
pi@raspberrypi:~/Coding/v8/out/arm.release $ ./mksnapshot --help
target arm v7 neon hard
ARMv8=0 ARMv7=1 VFPv3=1 VFP32DREGS=1 NEON=1 SUDIV=1 USE_EABI_HARDFLOAT=1`

**Comment by groundwater** [comment]
@hypersad that looks promising, does your `mksnapshot` work natively on ARM?

**Comment by hypersad** [comment]
@groundwater yep, I've generated snapshot but can't test it yesterday and gonna try it today. This way for building it is so weird, but at least we know, that is possible.

**Comment by hypersad** [comment]
Is anyone else working on this? Yep, we could compile armv7l binaries and they are working, but I can't figure out, why Atom throws `Illegal instruction` error when using startup blob compiled by armv7l. I think, there can be something else, that "miscompiled". Maybe I should try to cross-compile electron with, ehhh, custom libs for libchromiumcontent? Any advices?

**Comment by hypersad** [comment]
@groundwater you're right! I tried to generate startup blob from startup.js on my `amd64` machine **with `armv7l`** binaries, and it works! I successfully built & launched Atom on my armv7l machine.

**Comment by mathuin** [comment]
@hypersad you are my hero!  Can you tell me exactly how you did it?  I would love to try your process on my Chromebook!

**Comment by hypersad** [comment]
@mathuin you can find all information about building Atom for arm platforms [here](https://github.com/atom/atom/issues/7822).

**Comment by hypersad** [comment]
@groundwater also, I'm still working on mksnapshot binaries and looking for your way of building it, what args you're using and how you do it? :eyes:

**Comment by hypersad** [comment]
@groundwater working `mksnapshot` binary for `armv7l` that generates proper snapshot is ready! :sunglasses:

**Comment by groundwater** [comment]
Fantastic @hypersad. Bringing in @jkleinsc who will be looking at this soon.

**Comment by hypersad** [comment]
I hope Electron 2.0.0 will include native working `mksnapshot` binaries?

**Comment by jkleinsc** [comment]
@hypersad I'm working on it now.  If it is ready, we certainly will include it in 2.0.

**Comment by sofianguy** [comment]
Thank you for taking the time to report this issue and helping to make Electron better.

The version of Electron you reported this on has been superseded by newer releases. 

If you're still experiencing this issue in the latest version of Electron, please add a comment specifying the version you're testing with and any other new information that a maintainer trying to reproduce the issue should know.

I'm setting the `blocked/need-info` label for the above reasons.

Thanks in advance! Your help is appreciated.
